solr/solrj/src/test/org/apache/solr/client/solrj/impl/HttpClusterStateSSLTest.java

```diff
@@ -0,0 +1,106 @@
+
+  @BeforeClass
+  public static void setupClusterWithSSL() throws Exception {
+    System.setProperty("solr.storeBaseUrl", "true");
+    configureCluster(1)
+        .addConfig("conf", getFile("solrj").toPath().resolve("solr").resolve("configsets").resolve("streaming").resolve("conf"))
+        .configure();
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void testHttpClusterStateWithSSL() throws Exception {
+    URL url0 = cluster.getJettySolrRunner(0).getBaseUrl();
+    assertEquals("https", url0.getProtocol());
+
+    final String collectionId = "test-https-cluster-state";
+    final int numShards = 3;
+    final int rf = 2;
+    final int expectedReplicas = numShards * rf;
+    CollectionAdminRequest.createCollection(collectionId, "conf", numShards, rf)
+        .setPerReplicaState(false)
+        .process(cluster.getSolrClient());
+    cluster.waitForActiveCollection(collectionId, numShards, numShards * rf);
+
+    // verify the base_url is actually stored with https in it on the server-side
+    byte[] stateJsonBytes = cluster.getZkClient().getData(ZkStateReader.getCollectionPath(collectionId), null, null, true);
+    assertNotNull(stateJsonBytes);
+    Map<String, Object> replicasMap =
+        (Map<String, Object>) Utils.getObjectByPath(Utils.fromJSON(stateJsonBytes), false, "/" + collectionId + "/shards/shard1/replicas");
+    assertEquals(2, replicasMap.size());
+    for (Object replicaObj : replicasMap.values()) {
+      Map<String, Object> replicaData = (Map<String, Object>) replicaObj;
+      String baseUrl = (String) replicaData.get("base_url");
+      assertTrue(baseUrl != null && baseUrl.startsWith("https://"));
+    }
+
+    // verify the http derived cluster state (on the client side) agrees with what the server stored
+    try (CloudSolrClient httpBasedCloudSolrClient = new CloudSolrClient.Builder(Collections.singletonList(url0.toExternalForm())).build()) {
+      ClusterStateProvider csp = httpBasedCloudSolrClient.getClusterStateProvider();
+      assertTrue(csp instanceof HttpClusterStateProvider);
+      verifyUrlSchemeInClusterState(csp.getClusterState(), collectionId, expectedReplicas);
+    }
+
+    // http2
+    try (CloudHttp2SolrClient http2BasedClient = new CloudHttp2SolrClient.Builder(Collections.singletonList(url0.toExternalForm())).build()) {
+      ClusterStateProvider csp = http2BasedClient.getClusterStateProvider();
+      assertTrue(csp instanceof Http2ClusterStateProvider);
+      verifyUrlSchemeInClusterState(csp.getClusterState(), collectionId, expectedReplicas);
+    }
+
+    // Zk cluster state now
+    ClusterStateProvider csp = cluster.getSolrClient().getClusterStateProvider();
+    assertTrue(csp instanceof ZkClientClusterStateProvider);
+    verifyUrlSchemeInClusterState(csp.getClusterState(), collectionId, expectedReplicas);
+  }
+
+  private void verifyUrlSchemeInClusterState(final ClusterState cs, final String collectionId, final int expectedReplicas) {
```

Comment:
The `collectionId` is always `test-https-cluster-state`. Do you want to just remove this parameter and declare it as static?

Reply:
Same with `expectedReplicas` (6)